app/jobs/upload_users_job.rb

```diff
@@ -0,0 +1,24 @@
+class UploadUsersJob < ApplicationJob
+  USER_FIELDS = [:user_name, :last_name, :first_name, :id_number, :email].freeze
+
+  def self.on_complete_js(_status)
+    'window.location.reload.bind(window.location)'
+  end
+
+  def perform(user_class, data, encoding)
+    progress.total = data.lines.count
+    user_class.transaction do
+      MarkusCsv.parse(data, encoding: encoding, skip_blanks: true, row_sep: :auto) do |row|
+        user_data = {}
+        USER_FIELDS.each_with_index do |field, index|
+          row_index = EndUser::CSV_ORDER.index(field) || index
+          user_data[field] = row[row_index]&.strip
+        end
+        next if user_data[:user_name].blank? || user_data[:last_name].blank? || user_data[:first_name].blank?
+        user = user_class.find_or_initialize_by(user_data)
```

Comment:
Let's allow *updates* of user information through CSV upload, keyed by `user_name`. So if the CSV contains a row of data for a username that already exists, MarkUs should *update* the existing user with that name using the values in the row, even if they're different from what the user currently has.

So then here, do a `find_or_initialize_by` just on the user name, and then update with the returned objects with other attributes.

Also, please update the tests and documentation to be consistent with this updating behaviour.

Reply:
I have made this updating behaviour work. It was implemented a little differently than you described though. This is because using the `find_or_initialize_by` just on the user name doesn't work since the `after_initialize :set_display_name` prevents users from being initialized without a first or last name. 
I instead did a find_by user name or initialize by all user attributes. Let me know if you prefer a different approach to working around this.